x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CachedBlobContainerIndexInput.java

```diff
@@ -517,6 +523,22 @@ void releaseOnClose() {
                 final CacheFile currentCacheFile = cacheFile.getAndSet(null);
                 if (currentCacheFile != null) {
                     currentCacheFile.release(this);
+                    try {
```

Comment:
>I wish we could avoid doing something like this, because adding eviction listeners has an impact on file handles and we don't want to keep a FileChannel opened for the cache file if there's no more IndexInput opened on it.

It's true that with the current approach it's possible that the reference for the `CacheFile` can live longer than needed, i.e. all `CachedBlobContainerIndexInput` are closed but we still have room in the cache and a `RecoveryStateCacheFileEvictionListener` holds a reference.

> I think that this RecoveryStateCacheFileEvictionListener is just needed in case of IndexInput being all close and CacheFile not evicted yet, so that when it will be evicted we can still update the recovery file details for it?

That's correct. When we `IndexInput` is closed we don't know if there are other references pointing to the same underlying `CacheFile` so we cannot remove it from the `FileDetails`.

> I'm wondering if we could keep the CacheFile in the CachedFileDetail and check in the recovered() method if it's being evicted, and then remove it from the cacheFiles Set there?

There's a minor problem with that approach, if the `FileInfo#recovered()` is not consumed those `CacheFile` instances could live forever, but I guess that should be ok? I'm not sure how common is to have multiple `IndexInput` pointing to the same `CacheFile` triggering this scenario.

Reply:
Thanks for your response, you perfectly understood what I might despite the poor quality of my previous comment.



> There's a minor problem with that approach, if the `FileInfo#recovered()` is not consumed those `CacheFile` instances could live forever, but I guess that should be ok?

I thought about it too, and I think we can circumvent that by also cleaning up the `CacheFile` set every time a CacheFile is added/removed or recovered bytes computed?